bgpd/bgp_route.c

```diff
@@ -1752,18 +1752,22 @@ int subgroup_announce_check(struct bgp_node *rn, struct bgp_path_info *pi,
 	 * if
 	 * the peer (group) is configured to receive link-local nexthop
 	 * unchanged
-	 * and it is available in the prefix OR we're not reflecting the route
-	 * and
+	 * and it is available in the prefix OR we're not reflecting the route,
+	 * link-local nexthop address is valid and
 	 * the peer (group) to whom we're going to announce is on a shared
 	 * network
 	 * and this is either a self-originated route or the peer is EBGP.
+	 * By checking if nexthop LL address is valid we are sure that
+	 * we do not announce LL address as `::`.
 	 */
 	if (NEXTHOP_IS_V6) {
 		attr->mp_nexthop_len = BGP_ATTR_NHLEN_IPV6_GLOBAL;
 		if ((CHECK_FLAG(peer->af_flags[afi][safi],
 				PEER_FLAG_NEXTHOP_LOCAL_UNCHANGED)
 		     && IN6_IS_ADDR_LINKLOCAL(&attr->mp_nexthop_local))
-		    || (!reflect && peer->shared_network
+		    || (!reflect
+			&& IN6_IS_ADDR_LINKLOCAL(&peer->nexthop.v6_local)
```

Comment:
@ton31337 I think this still have issues. When length is not equal to "BGP_ATTR_NHLEN_IPV6_GLOBAL_AND_LL", it will not add any new LL address to the stream in function (), however if the stream already have :: as LL, then stream is un modified. isn't it ?  

Interested to know, how did you verify this ? Would be helpful if you have packet captures of BGP update packet for this case.

Reply:
@srimohans I added screenshots
with patch:
![Screenshot from 2019-09-11 14-07-07](https://user-images.githubusercontent.com/3352707/64692322-7fc3c080-d49d-11e9-963e-07e24f05e164.png)
without patch:
![Screenshot from 2019-09-11 14-06-11](https://user-images.githubusercontent.com/3352707/64692323-7fc3c080-d49d-11e9-9f03-6195c11d491a.png)

Reply:
pcaps:
http://donatas.net/before_validation.pcap
http://donatas.net/after_validation.pcap